src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,438 @@
+at infinity. -/
+class modular_form_class extends slash_invariant_form_class F Γ k :=
+(hol : ∀ f : F, mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (f : ℍ → ℂ))
+(bdd_at_infty : ∀ (f : F) (A : SL(2, ℤ)), is_bounded_at_im_infty (f ∣[k, A]))
+
+/--`cusp_form_class F Γ k` says that `F` is a type of bundled functions that extend
+`slash_invariant_forms_class` by requiring that the functions be holomorphic and zero
+at infinity. -/
+class cusp_form_class extends slash_invariant_form_class F Γ k :=
+(hol : ∀ f : F, mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (f : ℍ → ℂ))
+(zero_at_infty : ∀ (f : F) (A : SL(2, ℤ)), is_zero_at_im_infty (f ∣[k, A]))
+
+@[priority 100]
+instance modular_form_class.modular_form : modular_form_class (modular_form Γ k) Γ k :=
+{ coe := modular_form.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  slash_action_eq := modular_form.slash_action_eq',
+  hol := modular_form.hol',
+  bdd_at_infty := modular_form.bdd_at_infty' }
+
+@[priority 100]
+instance cusp_form_class.cusp_form : cusp_form_class (cusp_form Γ k) Γ k :=
+{ coe := cusp_form.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  slash_action_eq := cusp_form.slash_action_eq',
+  hol := cusp_form.hol',
+  zero_at_infty := cusp_form.zero_at_infty' }
+
+variables {F Γ k}
+
+@[simp] lemma modular_form_to_fun_eq_coe {f : modular_form Γ k} : f.to_fun = (f : ℍ → ℂ) := rfl
+@[simp] lemma cusp_form_to_fun_eq_coe {f : cusp_form Γ k} : f.to_fun = (f : ℍ → ℂ) := rfl
+
+@[ext] theorem modular_form.ext {f g : modular_form Γ k} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+@[ext] theorem cusp_form.ext {f g : cusp_form Γ k} (h : ∀ x, f x = g x) : f = g :=
+fun_like.ext f g h
+
+/-- Copy of a `modular_form` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def modular_form.copy (f : modular_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  modular_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',
+  hol' := h.symm ▸ f.hol',
+  bdd_at_infty' := λ A, h.symm ▸ f.bdd_at_infty' A }
+
+/-- Copy of a `cusp_form` with a new `to_fun` equal to the old one. Useful to fix
+definitional equalities. -/
+protected def cusp_form.copy (f : cusp_form Γ k) (f' : ℍ → ℂ) (h : f' = ⇑f) :
+  cusp_form Γ k :=
+{ to_fun := f',
+  slash_action_eq' := h.symm ▸ f.slash_action_eq',
+  hol' := h.symm ▸ f.hol',
+  zero_at_infty' := λ A, h.symm ▸ f.zero_at_infty' A }
+
+end modular_form
+
+namespace modular_form
+
+open slash_invariant_form
+
+variables {F : Type*} {Γ : subgroup SL(2, ℤ)} {k : ℤ}
+
+instance has_add : has_add (modular_form Γ k) :=
+{ add := λ f g,
+  { to_fun := f + g,
+    slash_action_eq' := (f.to_slash_invariant_form + g.to_slash_invariant_form).2,
+    hol' := f.hol'.add g.hol',
+    bdd_at_infty' := λ A, begin
+        rw slash_action.add_action,
+        exact (bounded_at_im_infty_subalgebra ℂ).add_mem' (f.bdd_at_infty' A) (g.bdd_at_infty' A),
+        end}}
```

Comment:
I suggest phrasing this proof in a way which explicitly inherits the well-defined addition on `slash_invariant_form`, rather than re-proving it:
```suggestion
{ add := λ f g,
  { hol' := f.hol'.add g.hol',
    bdd_at_infty' := λ A, begin
      convert (bounded_at_im_infty_subalgebra ℂ).add_mem (f.bdd_at_infty' A) (g.bdd_at_infty' A),
      simp [slash_add],
    end,
    .. (f : slash_invariant_form Γ k) + g } }
```
Same for `has_zero`, `has_csmul`, etc., and same throughout for `cusp_form`.